pageserver/src/tenant.rs

```diff
@@ -1370,9 +1389,23 @@ impl Tenant {
         timeline.walreceiver.stop().await;
         debug!("wal receiver shutdown confirmed");
 
+        if let Some(remote_client) = timeline.remote_client.as_ref() {
+            remote_client.stop()?;
+        }
+
+        // NOTE: there is no guarantee that after this point no tasks related to a timeline are running.
+        //     Tasks can spawn new tasks before they see the cancellation request.
+        //     There is nothing preventing spawning new tasks after shutdown is requested for particular timeline or tenant.
         info!("waiting for timeline tasks to shutdown");
         task_mgr::shutdown_tasks(None, Some(self.tenant_id), Some(timeline_id)).await;
 
+        // Make sure we marked timeline as deleted on the remote
+        // so we wont pick it up next time during attach or pageserver restart
+        // See comment in persist_index_part_with_deleted_flag.
+        if let Some(remote_client) = timeline.remote_client.as_ref() {
+            remote_client.persist_index_part_with_deleted_flag().await?;
+        }
+
         {
```

Comment:
Regarding that `NOTE`: I'm missing an explainer why it's safe to proceeed anyways.

Maybe it isn't safe, but, to be safe, we'd need the lifecycle rework, so, until we have it, we accept it's not safe?

Reply:
It is safe because we stop remote_client above. So no new uploads can happen even if tasks are running. 

To fix the behavior we indeed need the lifecycle rework

Reply:
I'll remove that `NOTE`, I think it's fair to say this is required implicit knowledge throughout the pageserver code base right now.

I added a comment on the remote_client.stop() call though, clarifying that we need to do that _before_ the shutdown_tasks call.